archive/imo/imo1964_q1.lean

```diff
@@ -0,0 +1,56 @@
+/-
+Copyright (c) 2020 Kenji Nakagawa. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Kenji Nakagawa
+-/
+
+import field_theory.finite.basic
+
+lemma modfact (n : ℕ) : (2^n : zmod 7) = 2 ^ (n : zmod 3).val :=
+begin
+  have h1 : ∀ (k : ℕ), k < 6 → (2^k : zmod 7) = 2 ^ (k : zmod 3).val := by dec_trivial,
+  haveI : fact (nat.prime 7) := by { delta fact, norm_num },
+  have h2 : (2 : zmod 7) ≠ 0 := dec_trivial,
+  rw ← nat.mod_add_div n 6,
+  simp only [h2, pow_add, pow_mul, bit0_zero, one_pow, add_zero, mul_one, zmod.cast_self,
+    zmod.pow_bit0, nat.cast_bit0, zero_mul, ne.def, nat.cast_add, not_false_iff, nat.cast_mul],
+  apply h1,
+  apply nat.mod_lt,
+  norm_num,
+end
+
+/-!
+# IMO 1964 Q1
+
+* (a) Find all positive integers `n` for which `2^n-1` is divisble by `7`.
+* (b) Prove that there is no positive integer `n` for which `2^n+1` is divisible by `7`.
+-/
+example (n : ℕ) (gt_zero : 0 < n) : ((2^n -1) : zmod 7) = 0 ↔ (n : zmod 3) = 0 :=
```

Comment:
I'm not exactly sure what our principles are for formalizing statements like this, but I think we could phrase this a smidge more literally, without causing any serious problems.

Would this be difficult?
```suggestion
example (n : ℤ) (gt_zero : 0 < n) : 7 | (2^n -1) ↔ 3 | n :=
```

Reply:
For the other proof it seems to be no problem to change this directly, but for this one, you need (\u(2^n-1 : nat) : zmod 7) = 2^n-1 which is giving me a lot of problems.

Reply:
For questions of the form "Find all numbers for which blah", I've been formalizing it as, first write a `problem_predicate` which is what the problem is giving, then write a `solution_predicate` for what you claim the solution is, in some sort of simple form, and then the main theorem is that the two are equal. If it's a pain to muck around with converting between modular arithmetic and regularly arithmetic, well basically yeah, that's one of the main difficulties with formalizing things. For any sort of IMO challenge we won't get the problems in the most convenient form, so it's better to describe the problem in the way most similar to the problem statement.